ext/objspace/objspace_dump.c

```diff
@@ -37,22 +40,104 @@ struct dump_config {
     unsigned int full_heap: 1;
 };
 
-PRINTF_ARGS(static void dump_append(struct dump_config *, const char *, ...), 2, 3);
 static void
-dump_append(struct dump_config *dc, const char *format, ...)
+dump_flush(struct dump_config *dc)
 {
-    va_list vl;
+    if (dc->buffer_len) {
+        if (dc->stream) {
+            fwrite(dc->buffer, 1, dc->buffer_len, dc->stream);
+        }
+        else if (dc->string) {
+            rb_str_cat(dc->string, dc->buffer, dc->buffer_len);
+        }
+        dc->buffer_len = 0;
+    }
+}
+
+static inline void
+buffer_inc_capa(struct dump_config *dc, unsigned long requested)
+{
+    unsigned long required;
+
+    if (requested + dc->buffer_len >= dc->buffer_capa) {
+        dump_flush(dc);
+    }
+
+    for (required = dc->buffer_capa; requested > required - dc->buffer_len; required <<= 1);
+
+    if (required > dc->buffer_capa) {
+        REALLOC_N(dc->buffer, char, required);
+        dc->buffer_capa = required;
+    }
+}
+
+static void buffer_append(struct dump_config *dc, const char *cstr, unsigned long len)
+{
+    if (len > 0) {
+        buffer_inc_capa(dc, len);
+        MEMCPY(dc->buffer + dc->buffer_len, cstr, char, len);
+        dc->buffer_len += len;
+    }
+}
+
+# define dump_append(dc, str) buffer_append(dc, (str), (long)strlen(str))
+
+PRINTF_ARGS(static void dump_appendf(struct dump_config *, const char *, ...), 2, 3);
+static void
+dump_appendf(struct dump_config *dc, const char *format, ...)
+{
+    va_list vl, vl2;
     va_start(vl, format);
+    va_copy(vl2, vl);
```

Comment:
@k0kubun @nobu I don't remember well.  Was it okay for us to use `va_copy`?  I think we once tried it then abandoned for some reason?

Reply:
For cross compiling? https://github.com/ruby/ruby/commit/0f0c32f24e9556cf3d605df2891841659e6293f6

Reply:
Interesting.

Since `vsnprintf` is now only used for numbers, I think I can either entirely replace it in favor of some sort of `itoa()` or simply ensure enough buffer capacity to avoid the need for calling it a second time, thus eliminating the `va_copy`.